### ISSUE TITLE
Thoughts about using prettier

### ISSUE THREAD

**Issue by aboyton** [airbnb/javascript#1548]
Many of us have been using [prettier](https://prettier.io) for a while now and have found it wonderful for formatting code.

Prettier obviously doesn't replace most of the lint rules in Airbnb's style guide (c.f. #1307), but it can definitely compliment it by freeing up a developer to not have to worry about code style (where should line breaks happen, how much should this be indented) and instead worry about more important things (like not using `for-of` #1122). The prettier website has a bunch of other people's reasons [why they like prettier](https://prettier.io/docs/en/why-prettier.html).

Obviously people can use prettier with Airbnb's style guide currently (as many of us have been), but I was wondering if Airbnb had considered making this the default.

**Comment by ljharb** [comment]
1. prettier only deals with code formatting, and there's a large number of things that eslint handles that prettier can't, so a linter will still always be required when developing (this isn't something you're claiming, but it's something i've seen many prettier users try to argue)
1. All code style, including all formatting, is something developers should worry about - the more you lean on autocorrect to fix your spelling, the less you understand about the language you're writing - this is true both in prose and in code. The goal should be to get you to write code that is compliant with the linter *on your first try* - something that comes from habit and repetition - not to let you mash your keyboard and hope the computer fixes it.
1. I've still not yet seen a concrete example of something prettier can do, that eslint could not be *made* to do - there probably are a number of things prettier can do that eslint currently can't, but that time should have been spent writing autofix PRs for eslint instead of developing a different tool.
1. Nobody's yet taken the time to concretely map out the differences between airbnb's guide, and prettier's formatting (with any necessary prettier options enabled, ofc)

While I'd be quite interested to read the research from the last item above, and would be very interested in hearing a concrete example from the third item — at this time Airbnb does not use prettier, and does not recommend its use.

**Comment by aboyton** [comment]
Thanks for the quick reply. Unfortunately this reply is not nearly so short 😢 

1. I definitely wasn't suggesting prettier would replace airbnb's style guide (as others have claimed and you realised that I wasn't), but that you could allow prettier be part of it. To quote from the [How does it compare to ESLint](https://github.com/prettier/prettier#how-does-it-compare-to-eslint-or-tslint-stylelint) section of their website:
> Linters have two categories of rules:
> 
> **Formatting rules**: eg: [max-len](http://eslint.org/docs/rules/max-len), [no-mixed-spaces-and-tabs](http://eslint.org/docs/rules/no-mixed-spaces-and-tabs), [keyword-spacing](http://eslint.org/docs/rules/keyword-spacing), [comma-style](http://eslint.org/docs/rules/comma-style)...
>
> Prettier alleviates the need for this whole category of rules! Prettier is going to reprint the entire program from scratch in a consistent way, so it's not possible for the programmer to make a mistake there anymore :)
>
> **Code-quality rules**: eg [no-unused-vars](http://eslint.org/docs/rules/no-unused-vars), [no-extra-bind](http://eslint.org/docs/rules/no-extra-bind), [no-implicit-globals](http://eslint.org/docs/rules/no-implicit-globals), [prefer-promise-reject-errors](http://eslint.org/docs/rules/prefer-promise-reject-errors)...
>
> Prettier does nothing to help with those kind of rules. They are also the most important ones provided by linters as they are likely to catch real bugs with your code!

Instead I'm simply recommending extending [`eslint-plugin-prettier`](https://github.com/prettier/eslint-plugin-prettier) and either manually disabling all the rules which conflict (or by using the list from [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier) but I'm guessing you'd prefer to be explicit rather than depending on another list).

---

2. We've found using Airbnb's style guide freed us as a company from having the internal arguments about style which was great. No longer did we need to waste brain power arguing about if we should [`object-shorthand`](http://eslint.org/docs/rules/object-shorthand.html) or [`dot-notation`](http://eslint.org/docs/rules/dot-notation.html), we could let you worry about that for us.

We've been using Airbnb's style guide with [`eslint-plugin-prettier`](https://github.com/prettier/eslint-plugin-prettier) and [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier) for a couple of months now on a few hundred thousands of code in our codebase and have found it to be a very similar experience of freeing the mind to worry about more important things. It's hard to describe, but not needing to worry about whitespace (which is really all prettier does) has allowed me to concentrate better on the actual code that I'm writing. I'd highly encourage you to try it. Many in the company were highly dubious at first but were quickly won over.

---

3. Technically there's absolutely _nothing_ that prettier can do that you can't do with ESLint, precisely because the lint rule [`eslint-plugin-prettier`](https://github.com/prettier/eslint-plugin-prettier) does exactly that. You can simply consider prettier as a souped up indentation and line break lint rule (it does a couple of extra things with parens).

That said, that's a bit of a cheats argument and I'm guessing not what you were after.

The indentation rule in ESLint has gone through massive rewrites in ESLint 4. That said, it still won't say that

```js
// good
foo(arg1, arg2, arg3, arg4);

// bad
foo(
  arg1,
  arg2,
  arg3,
  arg4,
);

// bad
foo(reallyLongArg(), omgSoManyParameters(), IShouldRefactorThis(), isThereSeriouslyAnotherOne());

// good
foo(
  reallyLongArg(),
  omgSoManyParameters(),
  IShouldRefactorThis(),
  isThereSeriouslyAnotherOne()
);
```

This is the trivial example often cited, but it goes much further, how do you break up ternaries on multiple lines, how do you indent arrow functions in big promise chains, etc. With prettier I don't need to remember, I just hit save and it reformats it in a way that looks pretty good.

 I used to think that developers should worry about all aspects of code style, but having tried prettier for a few months now I'm glad to now have to worry about it.

---

4. Let me try my best. Prettier tries really hard to have as few options as possible (there's a fork that adds more, but I'd argue that's missing a lot of the point of not worrying about style).

The only two options you'd need for prettier are `--single-quote` and `--trailing-comma all`.

It's hard for me to know exactly what is different between using prettier to manage "style" and airbnb's rules. One way to see the differences would be to study the source code of [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier).

Another way (what I've done below) is see the practical differences we have on our codebase between the two. Please note that we don't use JSX and so I can't speak of any differences there.

Running prettier over our codebase for the first time reformatted a lot of whitespace changes that Airbnb's style guide didn't enforce. As for actual differences (as in places where the two disagree), the main one we found in our codebase (by number of lines) is:
```js
// airbnb (kinda)
const foo = function (arg, arg2) {

// prettier
const foo = function(arg, arg2) {
```
That said, [7.1](https://github.com/airbnb/javascript#functions--declarations) of the airbnb style guide forbids function declarations like this, we just haven't cleaned all of ours up yet :cold_sweat:.

As for the actual differences, these are all that I have found:

```js
/// space-in-parens
// airbnb
for (let i = 0; i < foo.length;) {

// prettier (note extra space after the last `;`
for (let i = 0; i < foo.length; ) {
```

```js
// wrap-iife
// airbnb
(function init() {
  // Body
}());

// prettier
(function init() {
  // Body
})();
```

```js
// indent
// airbnb
const foo = bar
  ? ''
  : {
    property,
    anotherProperty,
  };

// prettier
const foo = bar
  ? ""
  : {
      property,
      anotherProperty
    };
```

```js
// indent
// airbnb
this.foo
  .getStuff(
  {
    bar
  },
    {},
  )
  .then()

// prettier
this.foo
  .getStuff(
    {
      bar
    },
    {}
  )
  .then();
```

One more that I will add is that is seems that using [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier) means that the following seems to be enforced. I haven't worked out exactly where that's coming from as it's not prettier as you can see from the [playground](https://prettier.io/playground/#%7B%22content%22%3A%22foobar%20%3D%20someList.reduce(%5Cn%20%20(foo%2C%20bar)%20%3D%3E%20%7B%20return%20bar%20%3D%3D%3D%20''%20%3F%20foo%20%3A%20bar%3B%20%7D%2C%5Cn%20%20''%2C%5Cn)%3B%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Atrue%2C%22trailingComma%22%3A%22all%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D)

```js
// no-confusing-arrow
// airbnb
foobar = someList.reduce(
  (foo, bar) => { return bar === '' ? foo : bar; },
  '',
);

// prettier with eslint-config-prettier?
foobar = someList.reduce(
  (foo, bar) => (bar === '' ? foo : bar),
  '',
);
```

So to summarise, the conflicts between the two style guides all seem very minor and I do think that adding prettier as an ESLint rule would actually work really well as part of Airbnb's style guide. That said, it's not my guide, so all I can do is ask you to consider it.

**Comment by ljharb** [comment]
I'd say the best use of anyone's time would be to PR any "missing" autofixes into eslint - with those, prettier serves no purpose, and there'd be no need to add it.

**Comment by rattrayalex** [comment]
@ljharb is that feasible with the current eslint autofix primitives? See eg [this comment](https://github.com/eslint/eslint/issues/8972#issuecomment-316838603) from the eslint repo owner. 

Given the core functionality of prettier - line length - I'm not sure that any eslint fix could be made in a form other than "run prettier and use it as an eslint rule", which is what eslint-plugin-prettier accomplishes. 

It might be worth your time to experiment with the tool; I think you may find it interesting. 

Perhaps Airbnb could fork prettier, and align its output with your own style. I imagine it would be popular.

**Comment by rattrayalex** [comment]
@ljharb if you haven't read the Haskell paper which outlines the core

**Comment by ljharb** [comment]
Forking without intent to upstream, nor without intent to make a significantly different product, isn't particularly good for the ecosystem.

That comment is from one (of the many) eslint maintainers (not an "owner"). Nothing about that comment suggests these things are impossible; on the contrary, it suggests that making those changes would be a much better use of time and effort than reinventing the wheel with a new tool.

If improving eslint beyond a certain point isn't feasible, so be it - but we're a) nowhere near that point, and b) a better tool would use *eslint configs*, the defacto standard for declaring code style, and not just impose a minimally configurable and arbitrary style rubric.

**Comment by rattrayalex** [comment]
I think a lot of people would love to see a tool with the power of prettier, the configurability of eslint, and the opinions of Airbnb. I know I would. 

You might do a little more research into the nature of prettier before claiming that its core functionality is possible with existing eslint tools, however.

**Comment by ljharb** [comment]
I believe all the research you're referring to is about what's possible, not what's impossible. I've yet to see any examples of something with prettier that's never possible with eslint; and if someone can provide one, then that'd be a great motivation to change eslint itself to have a similar approach as prettier does.

**Comment by rattrayalex** [comment]
Mostly I meant its implementation, not its capabilities. Any program can be built with any Turing-complete tool, but elegant architectures can get the job done better. 

I think you'd enjoy a read of its guts if you haven't given them a gander (I know you're busy, so this would go in the "optional pleasure reading" bucket). I'm sure you'd find the source approachable. 

As @aboyton mentioned above, the key breakthrough of prettier is expanding/collapsing long expressions across lines based ~purely on AST. I think many people in the industry would be quite impressed if you were able to solve this problem without primitives like those prettier uses.

**Comment by rattrayalex** [comment]
To be clear, I *do* agree with you that it'd be possible to use those primitives as individual eslint rules, and in some ways preferable. 

Iit'd just be very difficult to build. 

(It'd also be too slow for a formatOnSave editor integration (which, while not something you recommend, is a very popular feature)).

**Comment by aboyton** [comment]
Sorry for the delay, I've been away and a busy with work.

It's 100% possible to make ESLint rules format code like prettier does, but the only real way to do so is to reimplement prettier, or to use prettier as an ESLint rule (which is what `eslint-plugin-prettier` does).

To demo how much stricter prettier is with styles compared to standard ESLint rules I've run prettier on `react-dates` as a PoC. https://github.com/airbnb/react-dates/pull/809 As you can see it changed a lot of code to be much more consistent with styling.
